### PR TITLE
GitHub CI

### PR DESCRIPTION
**Pull request by cjyar**

### CODE REVIEW

**Comment by cjyar** [comment]
Some notes:
- I copied the build from gridiron and modified it.
- I combined the cross builds for all the architectures into one build matrix. It's not very matrixy, but it lets us specify those steps just once.
- I combined `Check` and `Test Suite` since they were nearly identical. Now we're running tests on Rust beta.
- I enabled running, not just compiling, bench. We don't have a dedicated machine to run the benchmarks on, so there'll be some variation in the bench results. We'll see how much over time.
- For PRs, bench will compare against base, which is typically master. For master commits, it'll compare against the previous revision.

**Comment by cjyar** [comment]
Problem:
```
Fiddle:recrypt-rs chris$ critcmp one two -t 5
group                        one                                    two
-----                        ---                                    ---
256-bit decrypt (level 0)    1.00      7.1±0.09ms        ? B/sec    1.07      7.6±0.61ms        ? B/sec
Fiddle:recrypt-rs chris$ echo $?
0
Fiddle:recrypt-rs chris$ critcmp one two -t 8
no benchmark comparisons to show
Fiddle:recrypt-rs chris$ echo $?
1
```

...in other words, critcmp will fail the build if the two benchmark runs are identical to within the configured threshold. I think this is the opposite of what we want, but also the threshold of 2% is way too low.

File: .github/workflows/ci.yaml [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+
+  test:
+    name: Test Suite
+    runs-on: ubuntu-18.04
+    strategy:
+      matrix:
+        rust:
+          - stable
+          - beta
+          - 1.38.0
```
Comment (clintfred) [inline .github/workflows/ci.yaml:49]: suggest: add comment about why we have pinned this version.
```suggestion
          - 1.38.0 #MSRV
```

File: .github/workflows/ci.yaml [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+          command: test
+
+  bench:
+    name: Bench compare
+    runs-on: ubuntu-18.04
+    steps:
+      - name: Find base revision
+        id: base
+        run: |
+          if [ "${{ github.base_ref }}" != "" ] ; then
```
Comment (clintfred) [inline .github/workflows/ci.yaml:68]: neat!

File: src/internal/pairing.rs [hunk 3]
```diff
@@ -335,11 +335,11 @@ impl PairingConfig for fp_480::Monty {
     fn naf_for_loop() -> Vec<i8> {
         // if comparing to recrypt-scala, the last two elements were removed manually instead of at runtime
         let mut r = vec![
-            0, 0, 0, -1, 0, 1, 0, 0, 0, -1, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, -1, 0,
-            1, 0, 1, 0, -1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 0, -1, 0,
-            0, -1, 0, 0, 0, 1, 0, 0, 0, 1, 0, -1, 0, 0, 0, 0, -1, 0, 0, 0, 0, 0, 0, 1, 0, 0, -1, 0,
-            0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, -1, 0, 0, 1, 0, 0, -1, 0, 0, 0, 0, 1, 0, 0, 0, -1, 0,
-            1, 0, -1, 0, 0, 1, 0//, 1, 0,
+            0, 0, 0, -1, 0, 1, 0, 0, 0, -1, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0,
```
Comment (clintfred) [inline src/internal/pairing.rs:338]: is this just a format change?

Comment (cjyar) [reply src/internal/pairing.rs:338]: Yep. It wouldn't pass the build without this change, because rustfmt is part of CI now. :)

**Pull request merged**